docs: improve example for mock API (#23759)
microsoft/playwright: docs/src/mock.md

    json.Add(new Dictionary<string, object> { { "name", "Playwright" }, { "id", 100 } });
    // Fulfill using the original response, while patching the response body
    // with the given JSON object.
    await route.FulfillAsync(new RouteFulfillResponse
    {
      Response = response,
      Json = json
    });
  }
);
// Go to the page
await page.GotoAsync("https://demo.playwright.dev/api-mocking");

// Assert that the Strawberry fruit is visible
await Expect(page.GetByTextAsync("Playwright", new() { Exact: true })).ToBeVisibleAsync();
```

```java
page.route("*/**/api/v1/fruits", route -> {
  Response response = route.fetch();
  Map<String, Object> json = response.json();
  json.add(new HashMap<String, Object> { { "name", "Playwright" }, { "id", 100 } });
  // Fulfill using the original response, while patching the response body
  // with the given JSON object.
  route.fulfill(new Route.FulfillOptions().setResponse(response).setJsonBody(json));
});

// Go to the page
page.goto("https://demo.playwright.dev/api-mocking");

// Assert that the Strawberry fruit is visible
assertThat(page.getByText("Playwright", new Page.GetByTextOptions().setExact(true))).isVisible();
```

> From the trace of our test we can see that the API was called and the response was modified.

Reviewer: ```suggestion
In the trace of our test we can see that the API was called and the response was modified.
```

![trace of test showing api being called and fulfilled](https://github.com/microsoft/playwright/assets/13063165/8b8dd82d-1b3e-428e-871b-840581fed439)

By inspecting the response we can see that our new fruit was added to the list.
![trace of test showing the mock response](https://github.com/microsoft/playwright/assets/13063165/03e6c87c-4ecc-47e8-9ca0-30fface25e9d)

Read more about [advanced networking](./network.md).

## Mocking with HAR files

> A HAR file is an HTTP Archive file that contains a record of all the network requests that are made when a page is loaded. It contains information about the request and response headers, cookies, content, timings, and more. You can use HAR files to mock network requests in your tests. You'll need to:

Reviewer: ```suggestion
A HAR file is an [HTTP Archive](http://www.softwareishard.com/blog/har-12-spec/) file that contains a record of all the network requests that are made when a page is loaded. It contains information about the request and response headers, cookies, content, timings, and more. You can use HAR files to mock network requests in your tests. You'll need to:
```

1. Record a HAR file.
1. Commit the HAR file alongside the tests.
1. Route requests using the saved HAR files in the tests.

### Recording a HAR file

> To record a HAR file we use the `routeFromHAR` method. This method takes in the path to the HAR file and an optional object of options.

Reviewer: ```suggestion
To record a HAR file we use [`method: Page.routeFromHAR`] or [`method: BrowserContext.routeFromHAR`] method. This method takes in the path to the HAR file and an optional object of options.
```

The options object can contain the URL so that only requests with the URL matching the specified glob pattern will be served from the HAR File. If not specified, all requests will be served from the HAR file.

> The `update` option updates the given HAR file with the actual network information or creates a new HAR file, in the hars folder, if one doesn't already exist. In order to record the HAR file, set `update` to true.

Reviewer: ```suggestion
Setting `update` option to true will create or update the HAR file with the actual network information instead of serving the requests from the HAR file. Use it when creating a test to populate the HAR with real data.
```

```js

> test('records or updates the HAR file', async ({

Reviewer: ```suggestion
test('records or updates the HAR file', async ({ page }) => {
```

  page,
}) => {

>   // Get the response and add to it

Reviewer: Wrong comment. Same in other snippets.

  await page.routeFromHAR('./hars/fruit.har', {
    url: '*/**/api/v1/fruits',
    update: true,
  });
  // Go to the page
  await page.goto('https://demo.playwright.dev/api-mocking');
  // Assert that the fruit is visible

>   await expect(page.getByText('Strawberry')).toBeVisible();

Reviewer: Don't need the assertion yet, while recording the har. It is also the wrong assertion, let's remove it. Same for other snippets.

  # Assert that the Playwright fruit is visible
  page.get_by_text("Strawberry").to_be_visible()
```

```csharp
await page.RouteFromHARAsync(
  "./hars/fruit.har",
  new RouteFromHAROptions
  {
    Url = "*/**/api/v1/fruits",
    Update = true,
  }
);

// Go to the page
await page.GotoAsync("https://demo.playwright.dev/api-mocking");

// Assert that the Playwright fruit is visible
await page.ExpectByTextAsync("Strawberry").ToBeVisibleAsync();
```

```java
page.routeFromHAR("./hars/fruit.har", new RouteFromHAROptions()
  .setUrl("*/**/api/v1/fruits")
  .setUpdate(true)
);

// Go to the page
page.goto("https://demo.playwright.dev/api-mocking");

// Assert that the Playwright fruit is visible
assertThat(page.getByText("Strawberry")).isVisible();
```

### Modifying a HAR file

Once you have recorded a HAR file you can modify it by opening the hashed .txt file inside your 'hars' folder and editing the JSON. This file should be committed to your source control. Anytime you run this test with `update: true` it will update your HAR file with the request from the API. 

```json
[
  {
    "name": "Playwright",
    "id": 100
  },
  // ... other fruits
]
```

### Replaying from HAR

> Use [`method: Page.routeFromHAR`] or [`method: BrowserContext.routeFromHAR`] to serve matching responses from the [HAR](http://www.softwareishard.com/blog/har-12-spec/) file. Set `update: false` as an option in the `routeFromHAR` method or simply remove this option as the default is false. This will run the test against the HAR file instead of hitting the API.

Reviewer: ```suggestion
Now that you have the HAR file recorded and modified the mock data, it can be used to serve matching responses in the test. For this, just turn off or simply remove the `update` option. This will run the test against the HAR file instead of hitting the API.
```

```js

> test('gets the json from HAR and checks the new fruit has been added', async ({

Reviewer: ```suggestion
test('gets the json from HAR and checks the new fruit has been added', async ({ page }) => {
```

  # Go to the page
  page.goto("https://demo.playwright.dev/api-mocking")

  # Assert that the Playwright fruit is visible
  page.get_by_text("Playwright", exact=True).to_be_visible()
```

```csharp
// Replay API requests from HAR.
// Either use a matching response from the HAR,
// or abort the request if nothing matches.
await page.RouteFromHARAsync(
  "./hars/fruit.har",
  new RouteFromHAROptions
  {
    Url = "*/**/api/v1/fruits",
    Update = false,
  }
);

// Go to the page
await page.GotoAsync("https://demo.playwright.dev/api-mocking");

// Assert that the Playwright fruit is visible
await page.ExpectByTextAsync("Playwright", new() { Exact = true }).ToBeVisibleAsync();
```

```java
// Replay API requests from HAR.
// Either use a matching response from the HAR,
// or abort the request if nothing matches.
page.routeFromHAR("./hars/fruit.har", new RouteFromHAROptions()
  .setUrl("*/**/api/v1/fruits")
  .setUpdate(false)
);

// Go to the page
page.goto("https://demo.playwright.dev/api-mocking");

// Assert that the Playwright fruit is visible
assertThat(page.getByText("Playwright", new Page.GetByTextOptions().setExact(true))).isVisible();
```

> From the trace of our test we can see that the route was fulfilled from the HAR file and the API was not called.

Reviewer: ```suggestion
In the trace of our test we can see that the route was fulfilled from the HAR file and the API was not called.
```

![trace showing the HAR file being used](https://github.com/microsoft/playwright/assets/13063165/1bd7ab66-ea4f-43c2-a4e5-ca17d4837ff1)

> If we inspect the response we can see our new fruit was added to the JSON, which was done by manually updating the hashed .txt file inside the 'hars' folder.

Reviewer: ```suggestion
If we inspect the response we can see our new fruit was added to the JSON, which was done by manually updating the hashed `.txt` file inside the `hars` folder.
```

![trace showing response from HAR file](https://github.com/microsoft/playwright/assets/13063165/db3117fc-7b02-4973-9a51-29e213261a6a)

HAR replay matches URL and HTTP method strictly. For POST requests, it also matches POST payloads strictly. If multiple recordings match a request, the one with the most matching headers is picked. An entry resulting in a redirect will be followed automatically.

Similar to when recording, if given HAR file name ends with `.zip`, it is considered an archive containing the HAR file along with network payloads stored as separate entries. You can also extract this archive, edit payloads or HAR log manually and point to the extracted har file. All the payloads will be resolved relative to the extracted har file on the file system.

> ### Recording HAR with CLI

Reviewer: ```suggestion
#### Recording HAR with CLI

We recommend the `update` option to record HAR file for your test. However, you can also record the HAR with Playwright CLI.
```

Open the browser with Playwright CLI and pass `--save-har` option to produce a HAR file. Optionally, use `--save-har-glob` to only save requests you are interested in, for example API endpoints. If the har file name ends with `.zip`, artifacts are written as separate files and are all compressed into a single `zip`.

```bash js
# Save API requests from example.com as "example.har" archive.
npx playwright open --save-har=example.har --save-har-glob="**/api/**" https://example.com
```

```bash java
# Save API requests from example.com as "example.har" archive.
mvn exec:java -e -D exec.mainClass=com.microsoft.playwright.CLI -D exec.args="open --save-har=example.har --save-har-glob='**/api/**' https://example.com"
```

```bash python
# Save API requests from example.com as "example.har" archive.
playwright open --save-har=example.har --save-har-glob="**/api/**" https://example.coms
```

```bash csharp
# Save API requests from example.com as "example.har" archive.
pwsh bin/Debug/netX/playwright.ps1 open --save-har=example.har --save-har-glob="**/api/**" https://example.com
```

> ### Recording HAR with a script

Reviewer: I'd just drop this section entirely.

[fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) requests, can be tracked, modified and mocked. With Playwright you can also mock using HAR files that contain multiple network requests made by the page.

## Mock API requests

The following code will intercept all the calls to `*/**/api/v1/fruits` and will return a custom response instead. No requests to the API will be made. The test goes to the URL that uses the mocked route and asserts that mock data is present on the page.

```js
test("mocks a fruit and doesn't call api", async ({ page }) => {
  // Mock the api call before navigating
  await page.route('*/**/api/v1/fruits', async (route) => {
    const json = [{ name: 'Strawberry', id: 21 }];
    await route.fulfill({ json });
  });
  // Go to the page
  await page.goto('https://demo.playwright.dev/api-mocking');
  
  // Assert that the Strawberry fruit is visible
  await expect(page.getByText('Strawberry')).toBeVisible();
});
```

```python async

> async def test_mock_the_fruit_api(page):

Reviewer 2: ```suggestion
async def test_mock_the_fruit_api(page: Page):
```

>     async def handle(route):

Reviewer 2: ```suggestion
    async def handle(route: Route):
```

>         json = [{name: "Strawberry", id: 21}]

Reviewer 2: ```suggestion
        json = [{"name": "Strawberry", "id": 21}]
```

        # fulfill the route with the mock data
        await route.fulfill(json=json)

    # Intercept the route to the fruit API
    await page.route("*/**/api/v1/fruits", handle)

    # Go to the page
    await page.goto("https://demo.playwright.dev/api-mocking")

    # Assert that the Strawberry fruit is visible
    await page.get_by_text("Strawberry").to_be_visible()
```

```python sync

> def test_mock_the_fruit_api(page):
>     def handle(route):
>         json = [{name: "Strawberry", id: 21}]
>         # fulfill the route with the mock data
>         route.fulfill(json=json)

Reviewer 2: ```suggestion
def test_mock_the_fruit_api(page: Page):
    def handle(route: Route):
        json = [{"name": "Strawberry", "id": 21}]
        # fulfill the route with the mock data
        route.fulfill(json=json)
```

    # Intercept the route to the fruit API
    page.route("*/**/api/v1/fruits", handle)

    # Go to the page
    page.goto("https://demo.playwright.dev/api-mocking")

    # Assert that the Strawberry fruit is visible
    page.get_by_text("Strawberry").to_be_visible()
```

```csharp
// Intercept the route to the fruit API

> await page.RouteAsync("*/**/api/v1/fruits", async route => {
>   var json = new Array [{ name: 'Strawberry', id: 21 }];
>   // fulfill the route with the mock data
>   await route.FulfillAsync(new RouteFulfillResponse
>   {
>     Json = json
>   });
> });

Reviewer 2: ```suggestion
await page.RouteAsync("*/**/api/v1/fruits", async route => {
  var json = new[] { new { name = "Strawberry", id = 21 } };
  // fulfill the route with the mock data
  await route.FulfillAsync(new()
  {
    Json = json
  });
});
```

// Go to the page
await page.GotoAsync("https://demo.playwright.dev/api-mocking");

// Assert that the Strawberry fruit is visible
await Expect(page.GetByTextAsync("Strawberry")).ToBeVisibleAsync();
```

```java
// Intercept the route to the fruit API
page.route("https://fruit.ceo/api/breeds/image/random", route -> {

>   Map<String, Object> json = new HashMap<>();
>   json = [{ name: 'Strawberry', id: 21 }];

Reviewer 2: ```suggestion
    List<Dictionary<String, Object>> data = new ArrayList<Dictionary<String, Object>>();
    Hashtable<String, Object> dict = new Hashtable<String, Object>();
    dict.put("name", "Strawberry");
    dict.put("id", 21);
    data.add(dict);
```

  // fulfill the route with the mock data

>   route.fulfill(new Route.FulfillOptions().setJsonBody(json));

Reviewer 2: ```suggestion
  route.fulfill(RequestOptions.create().setData(data));
```

});

// Go to the page

> page.goto("https://demo.playwright.dev/api-mocking");

Reviewer 2: ```suggestion
page.navigate("https://demo.playwright.dev/api-mocking");
```

// Assert that the Strawberry fruit is visible

> assertThat(page.getByText("Strawberry")).toBeVisible();

Reviewer 2: ```suggestion
assertThat(page.getByText("Strawberry")).isVisible();
```

```

You can see from the trace of the example test that the API was never called, it was however fulfilled with the mock data.
![api mocking trace](https://github.com/microsoft/playwright/assets/13063165/3dc14cbf-c100-4efc-ac21-d7b52d698b53)

Read more about [advanced networking](./network.md).

## Modify API responses

Sometimes, it is essential to make an API request, but the response needs to be patched to
allow for reproducible testing. In that case, instead of mocking the request, one
can perform the request and fulfill it with the modified response.

In the example below we intercept the call to the fruit API and add a new fruit called 'playwright', to the data. We then go to the url and assert that this data is there:

```js
test('gets the json from api and adds a new fruit', async ({ page }) => {
  // Get the response and add to it
  await page.route('*/**/api/v1/fruits', async (route) => {
    const response = await route.fetch();
    const json = await response.json();

>     json.push({ name: "Playwright", id: 100 });

Reviewer 2: ```suggestion
    json.push({ name: 'Playwright', id: 100 });
```

    // Fulfill using the original response, while patching the response body
    // with the given JSON object.
    await route.fulfill({ response, json });
  });

  // Go to the page
  await page.goto('https://demo.playwright.dev/api-mocking');

  // Assert that the new fruit is visible
  await expect(page.getByText('Playwright', { exact: true })).toBeVisible();
});
```

```python async

> async def test_gets_the_json_from_api_and_adds_a_new_fruit(page):
>     async def handle(route):
>         response = await route.fulfill()
>         json = await response.json()
>         json.append({name: "Playwright", id: 100})
>         # Fulfill using the original response, while patching the response body
>         # with the given JSON object.
>         await route.fulfill(response=response, json=json)
> 
>     await page.route("https://dog.ceo/api/breeds/list/all", handle)
> 
>     # Go to the page
>     await page.goto("https://demo.playwright.dev/api-mocking")

Reviewer 2: ```suggestion
async def test_gets_the_json_from_api_and_adds_a_new_fruit(page: Page):
    async def handle(route: Route):
        response = await route.fulfill()
        json = await response.json()
        json.append({ "name": "Playwright", "id": 100})
        # Fulfill using the original response, while patching the response body
        # with the given JSON object.
        await route.fulfill(response=response, json=json)

    await page.route("https://dog.ceo/api/breeds/list/all", handle)

    # Go to the page
    await page.goto("https://demo.playwright.dev/api-mocking")
```

    # Assert that the new fruit is visible
    await page.get_by_text("Playwright", exact=True).to_be_visible()
```

```python sync

> def test_gets_the_json_from_api_and_adds_a_new_fruit(page):
>     def handle(route):
>         response = route.fulfill()
>         json = response.json()
>         json.append({name: "Playwright", id: 100})
>         # Fulfill using the original response, while patching the response body
>         # with the given JSON object.
>         route.fulfill(response=response, json=json)

Reviewer 2: ```suggestion
def test_gets_the_json_from_api_and_adds_a_new_fruit(page: Page):
    def handle(route: Route):
        response = route.fulfill()
        json = response.json()
        json.append({ "name": "Playwright", "id": 100})
        # Fulfill using the original response, while patching the response body
        # with the given JSON object.
        route.fulfill(response=response, json=json)
```

    page.route("https://dog.ceo/api/breeds/list/all", handle)

    # Go to the page
    page.goto("https://demo.playwright.dev/api-mocking")

    # Assert that the new fruit is visible
    page.get_by_text("Playwright", exact=True).to_be_visible()
```

```csharp
await page.RouteAsync("*/**/api/v1/fruits", async (route) => {
    var response = await route.FetchAsync();

>     var json = await response.JsonAsync();
>     json.Add(new Dictionary<string, object> { { "name", "Playwright" }, { "id", 100 } });

Reviewer 2: ```suggestion
    var fruits = await response.JsonAsync<Fruit[]>();
    fruits.Add(new Fruit() { Name = "Playwright", Id = 100 });
```

    // Fulfill using the original response, while patching the response body
    // with the given JSON object.

>     await route.FulfillAsync(new RouteFulfillResponse

Reviewer 2: ```suggestion
    await route.FulfillAsync(new ()
```

    {
      Response = response,
      Json = json
    });
  }
);
// Go to the page
await page.GotoAsync("https://demo.playwright.dev/api-mocking");

// Assert that the Strawberry fruit is visible
await Expect(page.GetByTextAsync("Playwright", new() { Exact: true })).ToBeVisibleAsync();
```

```java
page.route("*/**/api/v1/fruits", route -> {
  Response response = route.fetch();

>   Map<String, Object> json = response.json();
>   json.add(new HashMap<String, Object> { { "name", "Playwright" }, { "id", 100 } });

Reviewer 2: ```suggestion
  byte[] json = response.body();
  parsed = new Gson().fromJson(json, JsonObject.class)
  parsed.add(new JsonObject().add("name", "Playwright").add("id", 100));
```

  // Fulfill using the original response, while patching the response body
  // with the given JSON object.

>   route.fulfill(new Route.FulfillOptions().setResponse(response).setJsonBody(json));

Reviewer 2: ```suggestion
  route.fulfill(new Route.FulfillOptions().setResponse(response).setBody(json.toString()));
```

1. Record a HAR file.
1. Commit the HAR file alongside the tests.
1. Route requests using the saved HAR files in the tests.

### Recording a HAR file

To record a HAR file we use [`method: Page.routeFromHAR`] or [`method: BrowserContext.routeFromHAR`] method. This method takes in the path to the HAR file and an optional object of options.
The options object can contain the URL so that only requests with the URL matching the specified glob pattern will be served from the HAR File. If not specified, all requests will be served from the HAR file.

Setting `update` option to true will create or update the HAR file with the actual network information instead of serving the requests from the HAR file. Use it when creating a test to populate the HAR with real data.

```js
test('records or updates the HAR file', async ({ page }) => {
  // Get the response from the HAR file
  await page.routeFromHAR('./hars/fruit.har', {
    url: '*/**/api/v1/fruits',
    update: true,
  });

  // Go to the page
  await page.goto('https://demo.playwright.dev/api-mocking');
});
```

```python async

> async def records_or_updates_the_har_file(page):

Reviewer 2: ```suggestion
async def records_or_updates_the_har_file(page: Page):
```

    # Get the response from the HAR file
    await page.route_from_har("./hars/fruit.har", url="*/**/api/v1/fruits", update=True)

    # Go to the page
    await page.goto("https://demo.playwright.dev/api-mocking")
```

```python sync

> def records_or_updates_the_har_file(page):

Reviewer 2: ```suggestion
def records_or_updates_the_har_file(page: Page):
```

    # Get the response from the HAR file
    page.route_from_har("./hars/fruit.har", url="*/**/api/v1/fruits", update=True)

    # Go to the page
    page.goto("https://demo.playwright.dev/api-mocking")
```

```csharp
// Get the response from the HAR file

> await page.RouteFromHARAsync("./hars/fruit.har",
>   new RouteFromHAROptions

Reviewer 2: ```suggestion
await page.RouteFromHARAsync("./hars/fruit.har", new ()
{
```

```json
[
  {
    "name": "Playwright",
    "id": 100
  },
  // ... other fruits
]
```

### Replaying from HAR

Now that you have the HAR file recorded and modified the mock data, it can be used to serve matching responses in the test. For this, just turn off or simply remove the `update` option. This will run the test against the HAR file instead of hitting the API.

```js
test('gets the json from HAR and checks the new fruit has been added', async ({ page }) => {
  // Replay API requests from HAR.
  // Either use a matching response from the HAR,
  // or abort the request if nothing matches.
  await page.routeFromHAR('./hars/fruit.har', {
    url: '*/**/api/v1/fruits',
    update: false,
  });

  // Go to the page
  await page.goto('https://demo.playwright.dev/api-mocking');

  // Assert that the Playwright fruit is visible
  await expect(page.getByText('Playwright', { exact: true })).toBeVisible();
});
```

```python async

> async def test_gets_the_json_from_har_and_checks_the_new_fruit_has_been_added(page):

Reviewer 2: ```suggestion
async def test_gets_the_json_from_har_and_checks_the_new_fruit_has_been_added(page: Page):
```

    # Replay API requests from HAR.
    # Either use a matching response from the HAR,
    # or abort the request if nothing matches.
    await page.route_from_har("./hars/fruit.har", url="*/**/api/v1/fruits", update=False)

    # Go to the page
    await page.goto("https://demo.playwright.dev/api-mocking")

    # Assert that the Playwright fruit is visible
    await page.get_by_text("Playwright", exact=True).to_be_visible()
```

```python sync

> def test_gets_the_json_from_har_and_checks_the_new_fruit_has_been_added(page):

Reviewer 2: ```suggestion
def test_gets_the_json_from_har_and_checks_the_new_fruit_has_been_added(page: Page):
```

    # Replay API requests from HAR.
    # Either use a matching response from the HAR,
    # or abort the request if nothing matches.
    page.route_from_har("./hars/fruit.har", url="*/**/api/v1/fruits", update=False)

    # Go to the page
    page.goto("https://demo.playwright.dev/api-mocking")

    # Assert that the Playwright fruit is visible
    page.get_by_text("Playwright", exact=True).to_be_visible()
```

```csharp
// Replay API requests from HAR.
// Either use a matching response from the HAR,
// or abort the request if nothing matches.

> await page.RouteFromHARAsync(
>   "./hars/fruit.har",
>   new RouteFromHAROptions

Reviewer 2: ```suggestion
await page.RouteFromHARAsync("./hars/fruit.har", new ()
```

title: "Mock APIs"
---

## Introduction

Web APIs are usually implemented as HTTP endpoints. Playwright provides APIs to **mock** and **modify** network traffic, both HTTP and HTTPS. Any requests that a page does, including [XHRs](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) and
[fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) requests, can be tracked, modified and mocked.

## Mock API requests

The following code will intercept all the calls to `https://dog.ceo/api/breeds/list/all` and will return
the test data instead. No requests to the `https://dog.ceo/api/breeds/list/all` endpoint will be made.

```js

> await page.route("https://dog.ceo/api/breeds/list/all", async (route) => {

Reviewer 2: ```suggestion
await page.route('https://dog.ceo/api/breeds/list/all', async (route) => {
```
we use single quotes everywhere.

  const json = {
    message: { playwright_dog: [] },
  };
  await route.fulfill({ json });
});
```

```python async
async def handle(route):
    json = { message: { "playwright_dog": [] } }
    await route.fulfill(json=json)

await page.route("https://dog.ceo/api/breeds/list/all", handle)
```

```python sync
async def handle(route):

>     json = { message: { "playwright_dog": [] } }

Reviewer 2: ```suggestion
    json = { "message": { "playwright_dog": [] } }
```

```csharp
await page.RouteAsync("https://dog.ceo/api/breeds/list/all", async route =>
{
    var json = new { message = new { playwright_dog = new List<string>() } };
    await route.FulfillAsync(new () { Json: json });
});
```

```java
page.route("https://dog.ceo/api/breeds/list/all", route -> {
  route.fulfill(new Route.FulfillOptions()
    .setBody("{\"message\":{\"playwright_dog\":[]}}"));
});
```

> In the example below we intercept the route to the dog API and return a custom response. We then go to that url and assert that the response is our mock data:

Reviewer: Why do we have both snippets above and snippets below? They seem to do the same thing? Let's have a single snippet and a good explanation for it.

Reviewer: Yeah, I meant keeping the full snippet and both explanations. Maybe highlight the `route` (with comments?) to focus the attention on the mocking part.

```js
test("mocks the dog api", async ({ page }) => {
  await page.route("https://dog.ceo/api/breeds/list/all", async (route) => {
    const json = {
      message: { playwright_dog: [] },
    };
    await route.fulfill({ json });
  });

>   await page.goto("https://dog.ceo/api/breeds/list/all");

Reviewer: How can we `goto` to the url that is a json api call? The page url must be different from the api url.

Author: because we dont have a real world example to use that has a frontend that they can see the results I thought it was a good idea to use the api url to go to as they can then see the changes direct in the json. Basically if a user copies that code snippet and runs it it will work exactly as it does which is really nice and easy for the user to see and understand. I find it super useful to visually see what is going on but open to suggestions

Reviewer: I understand the motivation. But I think it turns a real example into useless one. Now, as a user, I would think that I can only mock the url that I directly `goto` to, which is never the case. In practice, I never mock the html, so this is very misleading imo. Should we find another example then?

>   await page.route("https://dog.ceo/api/breeds/list/all", async (route) => {
>     const json = {
>       message: { playwright_dog: [] },
>     };
>     await route.fulfill({ json });
>   });
>   await page.goto("https://dog.ceo/api/breeds/list/all");
> 
>   await expect(page.getByText("playwright_dog")).toBeVisible();

Reviewer 2: ```suggestion
  await page.route('https://dog.ceo/api/breeds/list/all', async (route) => {
    const json = {
      message: { playwright_dog: [] },
    };
    await route.fulfill({ json });
  });
  await page.goto('https://dog.ceo/api/breeds/list/all');

  await expect(page.getByText('playwright_dog')).toBeVisible();
```

});
```

```python async
async def test_mock_the_dog_api(page):
    await page.route('https://dog.ceo/api/breeds/list/all', async (route) => {

>         json = { message: { "playwright_dog": [] } }

Reviewer 2: ```suggestion
        json = { "message": { "playwright_dog": [] } }
```

        await route.fulfill(json=json)
    })
    await page.goto('https://dog.ceo/api/breeds/list/all')

    await page.expect_to_be_visible('playwright_dog')
```

```python sync
def test_mock_the_dog_api(page):
    page.route('https://dog.ceo/api/breeds/list/all', lambda route: {

>         json = { message: { "playwright_dog": [] } }

Reviewer 2: ```suggestion
        json = { "message": { "playwright_dog": [] } }
```

        route.fulfill(json=json)
    })

>     page.goto('https://dog.ceo/api/breeds/list/all')

Reviewer 2: ```suggestion
    page.goto("https://dog.ceo/api/breeds/list/all")
```
(double quotes in Python)

>     page.expect_to_be_visible('playwright_dog')

Reviewer 2: ```suggestion
    page.expect_to_be_visible("playwright_dog")
```

```

```csharp
await page.RouteAsync("https://dog.ceo/api/breeds/list/all", async route =>
{
  var json = new { message = new { playwright_dog = new List<string>() } };
  await route.FulfillAsync(new RouteFulfillResponse
  {
    Json = json
  });
});

await page.GotoAsync("https://dog.ceo/api/breeds/list/all");

await Expect(page.GetByText("playwright_dog")).ToBeVisibleAsync();
```

```java
page.route("https://dog.ceo/api/breeds/list/all", route -> {
  var json = new HashMap<String, Object>();
  json.put("message", new HashMap<String, Object>() {{
    put("playwright_dog", new ArrayList<String>());
  }});
  route.fulfill(new Route.FulfillOptions()
    .setJsonBody(json));
});

page.goto("https://dog.ceo/api/breeds/list/all");

assertThat(page.getByText("playwright_dog")).isVisible();
```

Read more about [advanced networking](./network.md).

## Modify API responses

Sometimes, it is essential to make an API request, but the response needs to be patched to
allow for reproducible testing. In that case, instead of mocking the request, one
can perform the request and fulfill it with the modified response.

```js

> await page.route("https://dog.ceo/api/breeds/list/all", async (route) => {

Reviewer 2: ```suggestion
await page.route('https://dog.ceo/api/breeds/list/all', async (route) => {
```
time to uninstall or fix prettier configuration? :p

  const response = await route.fetch();
  const json = await response.json();

>   json.message["playwright_dog"] = [];

Reviewer 2: ```suggestion
  json.message['playwright_dog'] = [];
```

  // with the given JSON object.
  route.fulfill(new Route.FulfillOptions()
    .setResponse(response)
    .setBody(json.toString()));
});
```

In the example below we intercept the call to the dog API and add a new breed of dog, 'playwright_dog', to the data. We then go to the url and assert that this data is there:

```js
test("intercept the dog api and add some data to it", async ({ page }) => {
  await page.route("https://dog.ceo/api/breeds/list/all", async (route) => {
    const response = await route.fetch();
    const json = await response.json();
    json.message["playwright_dog"] = [];

    await route.fulfill({ response, json });
  });
  await page.goto("https://dog.ceo/api/breeds/list/all");

  await expect(page.getByText("playwright_dog")).toBeVisible();
});
```

```python async
async def test_intercept_the_dog_api_and_add_some_data_to_it(page):
    await page.route('https://dog.ceo/api/breeds/list/all', async (route) => {
        response = await route.fulfill()
        json = await response.json()
        json["message"]["playwright_dog"] = []

        await route.fulfill(response=response, json=json)
    })
    await page.goto('https://dog.ceo/api/breeds/list/all')

>     await page.expect_to_be_visible('playwright_dog')

Reviewer 2: ```suggestion
    await expect(page.get_by_text("playwright_dog")).to_be_visible()
```

```

```python sync
def test_intercept_the_dog_api_and_add_some_data_to_it(page):
    page.route('https://dog.ceo/api/breeds/list/all', lambda route: {
        response = route.fulfill()
        json = response.json()
        json["message"]["playwright_dog"] = []

        route.fulfill(response=response, json=json)
    })
    page.goto('https://dog.ceo/api/breeds/list/all')

>     page.expect_to_be_visible('playwright_dog')

Reviewer 2: ```suggestion
    expect(page.get_by_text("playwright_dog")).to_be_visible()
```
